src/UserSpaceInstrumentation/AccessTraceesMemoryTest.cpp

```diff
@@ -18,47 +21,191 @@
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Unable to open file"));
+
+  // Read 0 bytes.
+  result = ReadTraceesMemory(pid, address, 0);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Tried to read 0 bytes."));
+
+  // Read past the end of the mappings.
+  result = ReadTraceesMemory(pid, address, length + 1);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Only got"));
+
+  // Read from bad address.
+  result = ReadTraceesMemory(pid, 0, length);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(
+      absl::StrContains(result.error().message(), "Failed to read from memory file of process"));
+
+  // Detach and end child.
+  ASSERT_FALSE(DetachAndContinueProcess(pid).has_error());
+  kill(pid, SIGKILL);
+  waitpid(pid, NULL, 0);
+}
+
+TEST(AccessTraceesMemoryTest, WriteFailures) {
+  pid_t pid = fork();
+  ASSERT_TRUE(pid != -1);
+  if (pid == 0) {
+    // Child just runs an endless loop.
+    while (true) {
+    }
+  }
+
+  // Stop the child process using our tooling.
+  ASSERT_TRUE(AttachAndStopProcess(pid).has_value());
+
+  auto continues_range_or_error = GetFirstContinuesAdressRange(pid);
+  ASSERT_TRUE(continues_range_or_error.has_value());
+  const auto continues_range = continues_range_or_error.value();
+  const uint64_t address = continues_range.first;
+  const uint64_t length = continues_range.second - continues_range.first;
+
+  // Backup.
+  auto backup = ReadTraceesMemory(pid, address, length);
+  ASSERT_TRUE(backup.has_value());
+
+  // Good write.
+  std::vector<uint8_t> bytes(length, 0);
+  auto result = WriteTraceesMemory(pid, address, bytes);
+  EXPECT_FALSE(result.has_error());
+
+  // Process does not exist.
+  result = WriteTraceesMemory(-1, address, bytes);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Unable to open file"));
+
+  // Write 0 bytes.
+  result = WriteTraceesMemory(pid, address, std::vector<uint8_t>());
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Tried to write 0 bytes."));
+
+  // Read past the end of the mappings.
+  bytes.push_back(0);
+  result = WriteTraceesMemory(pid, address, bytes);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Only wrote"));
+
+  // Read from bad address.
+  result = WriteTraceesMemory(pid, 0, bytes);
+  ASSERT_TRUE(result.has_error());
+  EXPECT_TRUE(absl::StrContains(result.error().message(), "Failed to write to memory file"));
+
+  // Restore, detach and end child.
+  ASSERT_FALSE(WriteTraceesMemory(pid, address, backup.value()).has_error());
+  ASSERT_FALSE(DetachAndContinueProcess(pid).has_error());
+  kill(pid, SIGKILL);
+  waitpid(pid, NULL, 0);
+}
+
 TEST(AccessTraceesMemoryTest, ReadWriteRestore) {
   pid_t pid = fork();
-  CHECK(pid != -1);
+  ASSERT_TRUE(pid != -1);
```

Comment:
I feel like this and a few below were better as `CHECK` as they are not really part assessing what is being tested, rather than the setup for the test is working as expected.

Reply:
Yep, this came up before and is tracked here: http://b/185312970 
I changed it in the places I touched here; the other stuff will be handled in a followup PR.